### PR TITLE
Add episode 31

### PR DESCRIPTION
**Pull request by bitsondatadev**

### CODE REVIEW

File: _episodes/31.md [hunk 1]
```diff
@@ -0,0 +1,579 @@
...
+affect Trino. Read [the blog for more details](https://trino.io/blog/2021/12/13/log4shell-does-not-affect-trino.html).
+
+<p align="center">
+ <img align="center" width="50%" src="/assets/episode/31/log4shell.jpeg"/><br/>
+</p>
+
+## Concept of the month: ReplicaSets, Deployments, and Services
+
+In [the first installment of Trinetes](/episodes/24.html), we covered what 
+containerization is and why we use it. We covered the difference between tools
```
Comment (mosabua) [inline _episodes/31.md:78]: We talked about

File: _episodes/31.md [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+like docker-compose and container orchestration systems like Kubernetes (k8s).
+Finally, we covered the first k8s object called a [_pod_](https://kubernetes.io/docs/concepts/workloads/pods/).
```
Comment (mosabua) [inline _episodes/31.md:80]: find other word than "covered" again

File: _episodes/31.md [hunk 3]
```diff
@@ -0,0 +1,579 @@
...
+
+As a reminder, a pod is the basic unit of deployment in a k8s cluster. In this
+episode, we cover how to scale, deploy, and connect these Pods. If you are
```
Comment (mosabua) [inline _episodes/31.md:83]: pods

File: _episodes/31.md [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+
+In [the first installment of Trinetes](/episodes/24.html), we covered what 
+containerization is and why we use it. We covered the difference between tools
+like docker-compose and container orchestration systems like Kubernetes (k8s).
+Finally, we covered the first k8s object called a [_pod_](https://kubernetes.io/docs/concepts/workloads/pods/).
+
+As a reminder, a pod is the basic unit of deployment in a k8s cluster. In this
+episode, we cover how to scale, deploy, and connect these Pods. If you are 
+missing some context, you should review [the first installment of this series](/episodes/24.html). 
+After we cover some new k8s objects, we cover the steps
```
Comment (mosabua) [inline _episodes/31.md:85]: lets get rid of this sentence.. or reword to avoid cover cover

File: _episodes/31.md [hunk 5]
```diff
@@ -0,0 +1,579 @@
...
+containerization is and why we use it. We covered the difference between tools
+like docker-compose and container orchestration systems like Kubernetes (k8s).
+Finally, we covered the first k8s object called a [_pod_](https://kubernetes.io/docs/concepts/workloads/pods/).
+
+As a reminder, a pod is the basic unit of deployment in a k8s cluster. In this
+episode, we cover how to scale, deploy, and connect these Pods. If you are 
+missing some context, you should review [the first installment of this series](/episodes/24.html). 
+After we cover some new k8s objects, we cover the steps 
+
+### Replica Sets
```
Comment (mosabua) [inline _episodes/31.md:87]: Replica sets

File: _episodes/31.md [hunk 6]
```diff
@@ -0,0 +1,579 @@
...
+
+_Replicas_ make one or more instances based on the same pod definitions. In k8s,
+the object used to manage replication is a [_ReplicaSet_](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/).
+
+ReplicaSets provide high availability by making multiple instances based on a
```
Comment (mosabua) [inline _episodes/31.md:92]: reword .. making multiple instances sounds weird

File: _episodes/31.md [hunk 7]
```diff
@@ -0,0 +1,579 @@
...
+
+_Replicas_ make one or more instances based on the same pod definitions. In k8s,
+the object used to manage replication is a [_ReplicaSet_](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/).
+
+ReplicaSets provide high availability by making multiple instances based on a 
+pod definition in the k8s cluster. Kubernetes automatically brings up any failed
+pod instances that go down in a ReplicaSets based on the number of replicas you
+specify in the definition.
+
+Replication also enables load balancing IO traffic over multiple pods. You then
```
Comment (mosabua) [inline _episodes/31.md:97]: remove "then"

File: _episodes/31.md [hunk 8]
```diff
@@ -0,0 +1,579 @@
...
+gain the flexibility to scale up or down as traffic increases or decreases
+without any downtime.
+
+To scale the number of pods in a live ReplicaSet, you can update the `replicas` 
+value in the ReplicaSet definition file then running the following command to
```
Comment (mosabua) [inline _episodes/31.md:102]: file, then

File: _episodes/31.md [hunk 9]
```diff
@@ -0,0 +1,579 @@
...
+update it:
+
+```
+kubectl replace -f replicaset-definition.yml
+```
+
+You can also edit the live ReplicaSet without changing the local file:
```
Comment (mosabua) [inline _episodes/31.md:109]: Probably should decide if you always use the code name "ReplicaSet" or imho better just treat is as normal words "replica sets". And only use the camel case when it is specifically about code..

File: _episodes/31.md [hunk 10]
```diff
@@ -0,0 +1,579 @@
...
+
+```
+kubectl edit replicaset <replicaset-name>
+```
+
+Note: ReplicaSets is that their definitions are generally defined in a
```
Comment (mosabua) [inline _episodes/31.md:115]: sorry.. dont understand what you are trying to say

File: _episodes/31.md [hunk 11]
```diff
@@ -0,0 +1,579 @@
...
+
+You can also edit the live ReplicaSet without changing the local file:
+
+```
+kubectl edit replicaset <replicaset-name>
+```
+
+Note: ReplicaSets is that their definitions are generally defined in a 
+Deployment object which will be covered in a couple sections. If you look for 
+the
```
Comment (mosabua) [inline _episodes/31.md:117]: sentence ends

File: _episodes/31.md [hunk 12]
```diff
@@ -0,0 +1,579 @@
...
+
+```
+kubectl edit replicaset <replicaset-name>
+```
+
+Note: ReplicaSets is that their definitions are generally defined in a 
+Deployment object which will be covered in a couple sections. If you look for 
+the 
+
+### Labels and Selectors
```
Comment (mosabua) [inline _episodes/31.md:119]: Labels and selectors

File: _episodes/31.md [hunk 13]
```diff
@@ -0,0 +1,579 @@
...
+Kubernetes object have [labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/) 
+which are just key/value properties used to identify and dynamically group k8s
+objects. Labels should be meaningful and relevant to k8s users to easily 
+comprehend things like which application, version, component, and environment 
+certain objects belong to. Labels are shared across instances, and so they are 
+not unique.
+
+[Selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors)
+specify the grouping of instance to target set of objects when deploying or 
+applying other operations over these objects. For example, a ReplicaSet that
```
Comment (mosabua) [inline _episodes/31.md:130]: Sorry.. what do you mean here?

File: _episodes/31.md [hunk 14]
```diff
@@ -0,0 +1,579 @@
...
+specify the grouping of instance to target set of objects when deploying or 
+applying other operations over these objects. For example, a ReplicaSet that 
+identifies a set of pods with its selector, will create new pods in the case of
+a pod crashing. 
+
+
+### Deployments
+
+[Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/)
+objects allow you to now take a ReplicaSet, and perform actions on that set
```
Comment (mosabua) [inline _episodes/31.md:138]: remove "now"

Comment (mosabua) [reply _episodes/31.md:138]: replica set

File: _episodes/31.md [hunk 15]
```diff
@@ -0,0 +1,579 @@
...
+
+The best way to start making sense of these concepts is to look at the k8s
+configuration files.
+
+```
+helm template tcb trino/trino --version 0.3.0
+```
+
+Below is the generated deployment configuration, 
+`trino/templates/deployment-worker.yaml` but with comments that deliniate where
```
Comment (mosabua) [inline _episodes/31.md:154]: delineate

File: _episodes/31.md [hunk 16]
```diff
@@ -0,0 +1,579 @@
...
+different sections of the configuration are defining.
+```
```
Comment (mosabua) [inline _episodes/31.md:156]: empty line before code block

File: _episodes/31.md [hunk 17]
```diff
@@ -0,0 +1,579 @@
...
+            {}
+```
+
+### ConfigMap
+
+You may have noticed that the pods define volumes that are referring to an
+object called [ConfigMap](https://kubernetes.io/docs/concepts/configuration/configmap/).
+This is a way to store non-confidential data in the form of key-value pairs.
+
+This is how the Trino chart loads the [Trino configurations](https://trino.io/docs/current/installation/deployment.html#configuring-trino)
```
Comment (mosabua) [inline _episodes/31.md:226]: Change to no have "This is" twice

File: _episodes/31.md [hunk 18]
```diff
@@ -0,0 +1,579 @@
...
+```
+
+### ConfigMap
+
+You may have noticed that the pods define volumes that are referring to an
+object called [ConfigMap](https://kubernetes.io/docs/concepts/configuration/configmap/).
+This is a way to store non-confidential data in the form of key-value pairs.
+
+This is how the Trino chart loads the [Trino configurations](https://trino.io/docs/current/installation/deployment.html#configuring-trino) 
+in the `/etc/trino` directory on the containers. This ConfigMap file,
```
Comment (mosabua) [inline _episodes/31.md:227]: The ConfigMap file

File: _episodes/31.md [hunk 19]
```diff
@@ -0,0 +1,579 @@
...
+    query.max-memory-per-node=1GB
+    query.max-total-memory-per-node=2GB
+    memory.heap-headroom-per-node=1GB
+    discovery.uri=http://tcb-trino:8080
+
+  log.properties: |
+    io.trino=INFO
+```
+
+The only other ConfigMap actually defines the [catalog files](https://trino.io/docs/current/installation/deployment.html#catalog-properties)
```
Comment (mosabua) [inline _episodes/31.md:278]: drop "actually"

File: _episodes/31.md [hunk 20]
```diff
@@ -0,0 +1,579 @@
...
+in the `/etc/trino/catalog` folder. This ConfigMap only contains the two TPC
```
Comment (mosabua) [inline _episodes/31.md:279]: This ConfigMap only defines two catalogs. They expose the TPC-H and TPC-DS benchmark datasets.

File: _episodes/31.md [hunk 21]
```diff
@@ -0,0 +1,579 @@
...
+    connector.name=tpch
+    tpch.splits-per-node=4
+  tpcds.properties: |
+    connector.name=tpcds
+    tpcds.splits-per-node=4
+```
+
+### Networking
+
+Unlike in the Docker world, where a Docker container always has an ip address,
```
Comment (mosabua) [inline _episodes/31.md:304]: IP address .. here and everywhere in this post

Comment (mosabua) [reply _episodes/31.md:304]: This sentence really is wrong .. The difference is that docker runs on the host directly and you can expose the container. In k8s the pod runs in the cluster and networking is private..

File: _episodes/31.md [hunk 22]
```diff
@@ -0,0 +1,579 @@
...
+```
+
+### Networking
+
+Unlike in the Docker world, where a Docker container always has an ip address, 
+in the k8s world a pod has its own internal IP address. This is done by k8s 
+creating an internal private network which is exposed by the ip address of the 
+k8s node. 
+
+Although these ip addresses can be used to address pods internally, it's not a
```
Comment (mosabua) [inline _episodes/31.md:309]: Drop Although and simplify by separating into multiple sentences.

File: _episodes/31.md [hunk 23]
```diff
@@ -0,0 +1,579 @@
...
+### Networking
+
+Unlike in the Docker world, where a Docker container always has an ip address, 
+in the k8s world a pod has its own internal IP address. This is done by k8s 
+creating an internal private network which is exposed by the ip address of the 
+k8s node. 
+
+Although these ip addresses can be used to address pods internally, it's not a 
+good idea as these ip addresses are dynamic and subject to change upon 
+termination and recreation. For this, you will want to set up routing that
```
Comment (mosabua) [inline _episodes/31.md:311]: reword to avoid "will want to"

File: _episodes/31.md [hunk 24]
```diff
@@ -0,0 +1,579 @@
...
+in the k8s world a pod has its own internal IP address. This is done by k8s 
+creating an internal private network which is exposed by the ip address of the 
+k8s node. 
+
+Although these ip addresses can be used to address pods internally, it's not a 
+good idea as these ip addresses are dynamic and subject to change upon 
+termination and recreation. For this, you will want to set up routing that 
+handles addressing via pod name vs ip address.  
+
+When you have multiple k8s nodes, you will have multiple ip addresses set up for
```
Comment (mosabua) [inline _episodes/31.md:314]: avoid will

File: _episodes/31.md [hunk 25]
```diff
@@ -0,0 +1,579 @@
...
+the nodes. The routing software must be set up to handle the assignment of the 
+internal networks to each nodes to avoid conflicts across the cluster. This type
+of  functionality exists in cloud services to handle this type of routing, such
```
Comment (mosabua) [inline _episodes/31.md:317]: cloud services, such as ..

File: _episodes/31.md [hunk 26]
```diff
@@ -0,0 +1,579 @@
...
+as Amazon EKS, Google GKE, and Azure AKS.
+
+### Services
+
+[_Services_](https://kubernetes.io/docs/concepts/services-networking/service/) 
+establish connectivity between different pods and can also make pods available
```
Comment (mosabua) [inline _episodes/31.md:323]: pods, and can make

File: _episodes/31.md [hunk 27]
```diff
@@ -0,0 +1,579 @@
...
+supported cloud providers. We won't cover this one, but this is used when 
+setting launching in EKS using the `eksctl`.
+
+Here's a diagram of the ClusterIP networking between different ReplicaSets.
+<p align="center">
+ <img align="center" width="75%" height="100%" src="/assets/episode/31/clusterip.png"/><br/>
+ Source: https://www.udemy.com/course/learn-kubernetes/
+</p>
+
+While enabling a node port to establish connectivity to a specific ReplicaSet of
```
Comment (mosabua) [inline _episodes/31.md:344]: Cut up into multiple sentences

File: _episodes/31.md [hunk 28]
```diff
@@ -0,0 +1,579 @@
...
+<p align="center">
+ <img align="center" width="75%" height="100%" src="/assets/episode/31/clusterip.png"/><br/>
+ Source: https://www.udemy.com/course/learn-kubernetes/
+</p>
+
+While enabling a node port to establish connectivity to a specific ReplicaSet of 
+pod instances, it cannot make a generically accessible ip address for services 
+to communicate between one another.
+
+In our case, we want to make the coordinator node have an external ip address.
```
Comment (mosabua) [inline _episodes/31.md:348]: we configure an external IP address for the coordinator.

File: _episodes/31.md [hunk 29]
```diff
@@ -0,0 +1,579 @@
...
+The helm chart defines a ClusterIP service to accomplish this. Notice the
```
Comment (mosabua) [inline _episodes/31.md:349]: Helm

File: _episodes/31.md [hunk 30]
```diff
@@ -0,0 +1,579 @@
...
+the node from the pod.
+
+<p align="center">
+ <img align="center" width="75%" height="100%" src="/assets/episode/31/service.png"/><br/>
+ Source: https://www.udemy.com/course/learn-kubernetes/
+</p>
+
+There are three ports when setting up a NodePort. 
+
+1. TargetPort - is the port number on the pod itself, where the service forward to.
```
Comment (mosabua) [inline _episodes/31.md:389]: forwards

File: _episodes/31.md [hunk 31]
```diff
@@ -0,0 +1,579 @@
...
+2. Port - is the port used by the service. Which itself has an ip address
```
Comment (mosabua) [inline _episodes/31.md:390]: What is this for "Which itself has an IP address" ?

File: _episodes/31.md [hunk 32]
```diff
@@ -0,0 +1,579 @@
...
+3. NodePort - is the port that is exposed by the worker node and made available 
+externally. NodePorts can only be in the range of 30000 - 32767.
+
+The only required port to set is `port`. By default `targetPort` will be the
```
Comment (mosabua) [inline _episodes/31.md:394]: avoid "will"

File: _episodes/31.md [hunk 33]
```diff
@@ -0,0 +1,579 @@
...
+
+The only required port to set is `port`. By default `targetPort` will be the 
+same as `port` and nodePort will get automatically assigned a free port in the 
+allowed range. `ports` is also an array which is why the `-` char is used. 
+
+
+### Amazon EKS (Elastic Kubernetes Service)
+
+Amazon EKS is a managed container service to run and scale Kubernetes 
+applications in the cloud or on-premises. EKS manages the control plane for you
```
Comment (mosabua) [inline _episodes/31.md:402]: remove "on-premises" imho

File: _episodes/31.md [hunk 34]
```diff
@@ -0,0 +1,579 @@
...
+allowed range. `ports` is also an array which is why the `-` char is used. 
+
+
+### Amazon EKS (Elastic Kubernetes Service)
+
+Amazon EKS is a managed container service to run and scale Kubernetes 
+applications in the cloud or on-premises. EKS manages the control plane for you 
+so you only have to worry about deployment and nodes. Unlike with your own k8s
+cluster, you can't log into the control plane node in EKS, although you won't
+need to. You will only be able to access workers which are usually EC2 nodes.
```
Comment (mosabua) [inline _episodes/31.md:405]: avoid "will"

File: _episodes/31.md [hunk 35]
```diff
@@ -0,0 +1,579 @@
...
+> the community. Create a basic cluster in minutes with just one command.
+
+## Demo of the month: Deploy Trino k8s to Amazon EKS
+
+First, you'll need to install the following tools if you haven't done so already:
+* [eksctl](https://github.com/weaveworks/eksctl)
+* [kubectl](https://kubernetes.io/docs/tasks/tools/)
+* [helm](https://helm.sh/docs/intro/install/)
+
+Then you will need to add your IAM credentials to the
```
Comment (mosabua) [inline _episodes/31.md:427]: avoid "will"

File: _episodes/31.md [hunk 36]
```diff
@@ -0,0 +1,579 @@
...
+eksctl create cluster \
+ --name tcb-cluster \
+ --version 1.21 \
+ --region us-east-1 \
+ --nodegroup-name k8s-tcb-cluster \
+ --node-type t2.large \
+ --nodes 2
+```
+
+When you run this it will take 10-15 minutes to set up. You will start out by
```
Comment (mosabua) [inline _episodes/31.md:443]: will

File: _episodes/31.md [hunk 37]
```diff
@@ -0,0 +1,579 @@
...
+2021-12-16 02:00:17 [ℹ]  waiting for at least 2 node(s) to become ready in "k8s-tcb-cluster"
+2021-12-16 02:00:17 [ℹ]  nodegroup "k8s-tcb-cluster" has 2 node(s)
+2021-12-16 02:00:17 [ℹ]  node "ip-192-168-2-123.ec2.internal" is ready
+2021-12-16 02:00:17 [ℹ]  node "ip-192-168-55-167.ec2.internal" is ready
+2021-12-16 02:00:18 [ℹ]  kubectl command should work with "~/.kube/config", try 'kubectl get nodes'
+2021-12-16 02:00:18 [✔]  EKS cluster "tcb-cluster" in "us-east-1" region is ready
+```
+
+Take special note that `eksctl` overwrote your k8s configuration to point you to 
+the eks cluster instead of a local cluster. To test that you can connect, run:
```
Comment (mosabua) [inline _episodes/31.md:470]: EKS cluster ?

File: _episodes/31.md [hunk 38]
```diff
@@ -0,0 +1,579 @@
...
+
+```
+kubectl get nodes
+```
+
+You should see two nodes running. Now everything is simple. All you have to do
+to install Trino is reuse the helm chart that we used to locally deploy Trino.
```
Comment (mosabua) [inline _episodes/31.md:477]: Trino is to reuse the Helm

File: _episodes/31.md [hunk 39]
```diff
@@ -0,0 +1,579 @@
...
+Now, with the exact same command, you will deploy to EKS since the tool updated
```
Comment (mosabua) [inline _episodes/31.md:478]: will

File: _episodes/31.md [hunk 40]
```diff
@@ -0,0 +1,579 @@
...
+your settings.
+
+```
+helm install tcb trino/trino --version 0.3.0
+```
+
+After you've installed the helm chart, wait a minute or two for the Trino
```
Comment (mosabua) [inline _episodes/31.md:485]: Helm

File: _episodes/31.md [hunk 41]
```diff
@@ -0,0 +1,579 @@
...
+
+You should see the output that the coordinator and both workers are available.
+
+```
+NAME                    READY   UP-TO-DATE   AVAILABLE   AGE
+tcb-trino-coordinator   1/1     1            1           67s
+tcb-trino-worker        2/2     2            2           67s
+```
+
+To connect to the cluster, the helm output gives pretty good instructions on how
```
Comment (mosabua) [inline _episodes/31.md:500]: Helm

File: _episodes/31.md [hunk 42]
```diff
@@ -0,0 +1,579 @@
...
+to create a tunnel from the cluster to your local laptop.
+
+```
+Get the application URL by running these commands:
+  export POD_NAME=$(kubectl get pods --namespace default -l "app=trino,release=tcb,component=coordinator" -o jsonpath="{.items[0].metadata.name}")
+  echo "Visit http://127.0.0.1:8080 to use your application"
+  kubectl port-forward $POD_NAME 8080:8080
+```
+
+Run that then go to <http://127.0.0.1:8080>, and you should see the Trino UI.
```
Comment (mosabua) [inline _episodes/31.md:510]: Run that, then go

File: _episodes/31.md [hunk 43]
```diff
@@ -0,0 +1,579 @@
...
+
+To clear out the helm install, run:
```
Comment (mosabua) [inline _episodes/31.md:512]: Helm

File: _episodes/31.md [hunk 44]
```diff
@@ -0,0 +1,579 @@
...
+
+```
+eksctl delete cluster --name test-cluster --region us-east-1
+```
+
+## PR of the month: 
+
+This weeks [PR of the month](https://github.com/trinodb/trino/issues/8921)
+implements `TRUNCATE TABLE`. This command is very similar to `DELETE` statements,
+with the exception that it does not log all the row deletions. This ends up
```
Comment (mosabua) [inline _episodes/31.md:529]: that it does not perform deletes on individual rows.

**Review by mosabua — CHANGES_REQUESTED** [review]
Gave it a first pass ..

File: _episodes/31.md (hunk 5 above)
Comment (bitsondatadev) [reply _episodes/31.md:93]: actually should be ReplicaSets

File: _episodes/31.md (hunk 9 above)
Comment (bitsondatadev) [reply _episodes/31.md:115]: I'll go with always using it for simplicity.

File: _episodes/31.md (hunk 10 above)
Comment (bitsondatadev) [reply _episodes/31.md:121]: Removing this...I state what I was trying to state here later. That a ReplicaSet definition is defined inside of a Deployment definition.

File: _episodes/31.md (hunk 33 above)
Comment (bitsondatadev) [reply _episodes/31.md:408]: I like this to drive the point home that workers are independent of the control plane and when you use EKS, or any of the other services, that is really the service being "cloudified".

Comment (mosabua) [reply _episodes/31.md:408]: While that is theoretically true in the real world that just doesnt hold up unless you have very specific circumstances and knowledge. The service and the cluster have to be as much colocated as possible to have any reasonable performance. The whole point of using EKS is to make it simple .. with trying to hook in external machines on prem or wherever you are making it very complicated... its an edge case usage at best and I would leave it out to avoid confusion .

File: _episodes/31.md [hunk 45]
```diff
@@ -0,0 +1,584 @@
...
+
+You can also edit the live ReplicaSet without changing the local file:
+
+```
+kubectl edit replicaset <replicaset-name>
+```
+
+### Labels and selectors
+
+Kubernetes object have [labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/)
```
Comment (mosabua) [inline _episodes/31.md:122]: objects

File: _episodes/31.md (hunk 13 above)
Comment (mosabua) [reply _episodes/31.md:131]: This is still not a sentence.. what are you trying to say?

File: _episodes/31.md [hunk 46]
```diff
@@ -0,0 +1,584 @@
...
+comprehend things like which application, version, component, and environment 
+certain objects belong to. Labels are shared across instances, and so they are 
+not unique.
+
+[Selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors)
+specify the grouping of instance to target set of objects when deploying or 
+applying other operations over these objects. For example, a ReplicaSet that 
+identifies a set of pods with its selector to manage. When creating the 
+ReplicaSet, k8s creates new pods defined in the ReplicaSet's selector 
+definition. If the pods crash, k8s brings up new pods and associate the new
```
Comment (mosabua) [inline _episodes/31.md:134]: associates

File: _episodes/31.md [hunk 47]
```diff
@@ -0,0 +1,584 @@
...
+    connector.name=tpch
+    tpch.splits-per-node=4
+  tpcds.properties: |
+    connector.name=tpcds
+    tpcds.splits-per-node=4
+```
+
+### Networking
+
+Unlike in the Docker world, where it runs on the host directly where you can
```
Comment (mosabua) [inline _episodes/31.md:308]: those first two sentences need to either be combined or rephrased

File: _episodes/31.md [hunk 48]
```diff
@@ -0,0 +1,584 @@
...
+    tpch.splits-per-node=4
+  tpcds.properties: |
+    connector.name=tpcds
+    tpcds.splits-per-node=4
+```
+
+### Networking
+
+Unlike in the Docker world, where it runs on the host directly where you can 
+expose the container. In k8s the pod runs in the cluster and networking is
```
Comment (mosabua) [inline _episodes/31.md:309]: cluster, and

File: _episodes/31.md [hunk 49]
```diff
@@ -0,0 +1,584 @@
...
+    connector.name=tpcds
+    tpcds.splits-per-node=4
+```
+
+### Networking
+
+Unlike in the Docker world, where it runs on the host directly where you can 
+expose the container. In k8s the pod runs in the cluster and networking is 
+private. This is done by k8s creating an internal private network which is 
+exposed by the IP address of the
```
Comment (mosabua) [inline _episodes/31.md:311]: from "which is" this is either wrong or not clear what you are trying to say

File: _episodes/31.md [hunk 50]
```diff
@@ -0,0 +1,584 @@
...
+### Services
+
+[_Services_](https://kubernetes.io/docs/concepts/services-networking/service/) 
+establish connectivity between different pods and can make pods available 
+from the external k8s node IP address. This enables loose coupling between 
+microservices in applications.
+
+The above example is showing a NodePort service. There are three service types.
+
+1. ClusterIP - the service creates a virtual ip inside the cluster to enable
```
Comment (mosabua) [inline _episodes/31.md:334]: IP

File: _episodes/31.md [hunk 51]
```diff
@@ -0,0 +1,584 @@
...
+establish connectivity between different pods and can make pods available 
+from the external k8s node IP address. This enables loose coupling between 
+microservices in applications.
+
+The above example is showing a NodePort service. There are three service types.
+
+1. ClusterIP - the service creates a virtual ip inside the cluster to enable 
+communication between different services. This service is the default when you
+don't specify a `type` value under `spec` in the configuration.
+2. NodePort - is used to expose the internal address of a pod using the ip
```
Comment (mosabua) [inline _episodes/31.md:337]: IP

File: _episodes/31.md [hunk 52]
```diff
@@ -0,0 +1,584 @@
...
+The above example is showing a NodePort service. There are three service types.
+
+1. ClusterIP - the service creates a virtual ip inside the cluster to enable 
+communication between different services. This service is the default when you
+don't specify a `type` value under `spec` in the configuration.
+2. NodePort - is used to expose the internal address of a pod using the ip 
+address and port of the node it is running on.
+3. Load Balancer - this service creates a load balancer for the application in 
+supported cloud providers. We won't cover this one, but this is used when 
+setting launching in EKS using the `eksctl`.
```
Comment (mosabua) [inline _episodes/31.md:341]: setting launching ??

File: _episodes/31.md [hunk 53]
```diff
@@ -0,0 +1,584 @@
...
+
+## Demo of the month: Deploy Trino k8s to Amazon EKS
+
+First, you'll need to install the following tools if you haven't done so already:
+* [eksctl](https://github.com/weaveworks/eksctl)
+* [kubectl](https://kubernetes.io/docs/tasks/tools/)
+* [helm](https://helm.sh/docs/intro/install/)
+
+Then you need to add your IAM credentials to the 
+[`~.aws/credentials` file](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html#cli-configure-files-where).
```
Comment (mosabua) [inline _episodes/31.md:433]: should be `~/.aws/credentials`

File: _episodes/31.md [hunk 54]
```diff
@@ -0,0 +1,584 @@
...
+eksctl create cluster \
+ --name tcb-cluster \
+ --version 1.21 \
+ --region us-east-1 \
+ --nodegroup-name k8s-tcb-cluster \
+ --node-type t2.large \
+ --nodes 2
+```
+
+When you run this takes 10-15 minutes to set up. This is the first output you
```
Comment (mosabua) [inline _episodes/31.md:448]: The command completes in 10 to 15 minutes.

**Review by mosabua — CHANGES_REQUESTED** [review]

File: _episodes/31.md [hunk 55]
```diff
@@ -0,0 +1,583 @@
...
+
+Kubernetes objects have [labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/) 
+which are just key/value properties used to identify and dynamically group k8s
+objects. Labels should be meaningful and relevant to k8s users to easily 
+comprehend things like which application, version, component, and environment 
+certain objects belong to. Labels are shared across instances, and so they are 
+not unique.
+
+[Selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors)
+specify the grouping of instance to target set of objects when deploying or
```
Comment (mosabua) [inline _episodes/31.md:130]: to target a set of objects

File: _episodes/31.md (hunk 13 above)
Comment (mosabua) [reply _episodes/31.md:131]: Still cant make sense of that sentence. Please reword.

File: _episodes/31.md [hunk 56]
```diff
@@ -0,0 +1,583 @@
...
+
+The best way to start making sense of these concepts is to look at the k8s
+configuration files.
+
+```
+helm template tcb trino/trino --version 0.3.0
+```
+
+Below is the generated deployment configuration, 
+`trino/templates/deployment-worker.yaml` but with comments that delineate where
```
Comment (mosabua) [inline _episodes/31.md:157]: either add , before "but" 

or remove "but

File: _episodes/31.md [hunk 57]
```diff
@@ -0,0 +1,583 @@
...
+### ConfigMap
+
+You may have noticed that the pods define volumes that are referring to an
+object called [_ConfigMap_](https://kubernetes.io/docs/concepts/configuration/configmap/).
+This is a way to store non-confidential data in the form of key-value pairs.
+
+ConfigMaps are how the Trino chart loads the [Trino configurations](https://trino.io/docs/current/installation/deployment.html#configuring-trino) 
+in the `/etc/trino` directory on the containers. The ConfigMap file, 
+`trino/templates/configmap-worker.yaml`, defines the files loaded into the 
+worker nodes. The only real difference between the worker ConfigMap is in the
```
Comment (mosabua) [inline _episodes/31.md:233]: The only real difference of the ConfigMap is ..

File: _episodes/31.md [hunk 58]
```diff
@@ -0,0 +1,583 @@
...
+    query.max-memory-per-node=1GB
+    query.max-total-memory-per-node=2GB
+    memory.heap-headroom-per-node=1GB
+    discovery.uri=http://tcb-trino:8080
+
+  log.properties: |
+    io.trino=INFO
+```
+
+The only other ConfigMap defines the [catalog files](https://trino.io/docs/current/installation/deployment.html#catalog-properties)
```
Comment (mosabua) [inline _episodes/31.md:282]: catalog properties files

File: _episodes/31.md (hunk 33 above)
Comment (mosabua) [reply _episodes/31.md:406]: So ... what are doing here? I dont think EKS is used much or at all with nodes outside of EKS . performance would be terrible. I would rephrase this section to say that EKS provides K8s clusters in the cloud for you without your having to manage the whole k8s services and platform .. which is a HUGE win. Running k8s yourself is really hard.

File: _episodes/31.md [hunk 59]
```diff
@@ -0,0 +1,583 @@
...
+allowed range. `ports` is also an array which is why the `-` char is used. 
+
+
+### Amazon EKS (Elastic Kubernetes Service)
+
+Amazon EKS is a managed container service to run and scale Kubernetes 
+applications in the cloud or on-premises. EKS manages the control plane for you 
+so you only have to worry about deployment and nodes. Unlike with your own k8s
+cluster, you can't log into the control plane node in EKS, although you won't
+need to. You are able to access workers which are usually EC2 nodes.
```
Comment (mosabua) [inline _episodes/31.md:409]: The workers are just pods in the cluster .. what are you trying to say here?

File: _episodes/31.md [hunk 60]
```diff
@@ -0,0 +1,583 @@
...
+ --name tcb-cluster \
+ --version 1.21 \
+ --region us-east-1 \
+ --nodegroup-name k8s-tcb-cluster \
+ --node-type t2.large \
+ --nodes 2
+```
+
+The command completed in 10 to 15 minutes. This is the first output you
+see.
```
Comment (mosabua) [inline _episodes/31.md:448]: see:

**Review by mosabua — CHANGES_REQUESTED** [review]

File: _episodes/31.md (hunk 59 above)
Comment (bitsondatadev) [reply _episodes/31.md:409]: I'm actually a little confused by their question. Workers run pods, they are not pods. I'm trying to convey that workers are simply nodes, and in AWS, those nodes are EC2 nodes.

**Review by mosabua — APPROVED** [review]
Ship it!
